### PR TITLE
[libc++] Fix triviality of std::pair for trivially copyable types without an assignment operator

### LINKED ISSUE #95428: [libc++] ABI break between Clang 15 and Clang 16 for some trivially copiable pairs in C++ >= 23
```c++
#include <type_traits>
#include <utility>

struct trivially_copyable_no_assignment {
  int arr[4];
  trivially_copyable_no_assignment& operator=(const trivially_copyable_no_assignment&) = delete;
};

static_assert(std::is_trivially_copyable<std::pair<trivially_copyable_no_assignment, int> >::value, "");

int main() { }
```

`is_trivially_copyable` was true on Clang 15 and became false on Clang 16 and above. This only affects C++23 and  C++26, which also means that we now have a different ABI between C++20 and C++23 in recent Clangs.

Godbolt: https://godbolt.org/z/hEWq59oeP

### PR DESCRIPTION
**Pull request by ldionne**
Since 83ead2b, std::pair would not be trivially copyable when it holds a trivially copyable type without an assignment operator. That is because pair gained an elligible copy-assignment-operator (the const version) in 83ead2b in C++ >= 23.

This means that the trivially copyable property of std::pair for such types would be inconsistent between C++11/14/17/20 (trivially copyable) and C++23/26 (not trivially copyable). This patch makes std::pair's behavior consistent in all Standard modes EXCEPT C++03, which is a pre-existing condition and we have no way of changing (also, it shouldn't matter because the std::is_trivially_copyable trait was introduced in C++11).

While this is not technically an ABI break, in practice we do know that folks sometimes use a different representation based on whether a type is trivially copyable. So we're treating 83ead2b as an ABI break and this patch is fixing said breakage.

This patch also adds tests stolen from #89652 that pin down the ABI of std::pair with respect to being trivially copyable.

Fixes #95428

### CODE REVIEW

**Comment by h-vetinari** [comment]
> While this is not technically an ABI break, in practice we do know that folks sometimes use a different representation based on whether a type is trivially copyable. So we're treating [83ead2b](https://github.com/llvm/llvm-project/commit/83ead2bbc5e14ca1beb776c062ebc36c38e8bb1c) as an ABI break and this patch is fixing said breakage.

Should this then also target 18.1.8 (if it materializes)?

**Comment by ldionne** [comment]
> Should this then also target 18.1.8 (if it materializes)?

It would make sense, yes. TBH I am scared of landing these kinds of changes without any baking time in the tree though.

File: libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/abi.trivially_copyable.compile.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+struct trivially_copyable_no_construction {
+  int arr[4];
+  trivially_copyable_no_construction()                                                     = default;
+  trivially_copyable_no_construction(const trivially_copyable_no_construction&)            = delete;
+  trivially_copyable_no_construction& operator=(const trivially_copyable_no_construction&) = default;
+};
+static_assert(std::is_trivially_copyable<trivially_copyable_no_construction>::value, "");
+
+static_assert((!std::is_trivially_copyable<std::pair<int&, int> >::value), "");
```
Comment (philnik777) [inline libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/abi.trivially_copyable.compile.pass.cpp:40]: Why the extra layer of parens?

File: libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/abi.trivially_copyable.compile.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+struct trivially_copyable {
+  int arr[4];
+};
+
+struct trivially_copyable_no_assignment {
+  int arr[4];
+  trivially_copyable_no_assignment& operator=(const trivially_copyable_no_assignment&) = delete;
+};
+static_assert(std::is_trivially_copyable<trivially_copyable_no_assignment>::value, "");
```
Comment (philnik777) [inline libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/abi.trivially_copyable.compile.pass.cpp:30]: We should probably also add a test for the move assignment.

**Review by philnik777 — APPROVED** [review]
LGTM with comments addressed.

File: libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/abi.trivially_copyable.compile.pass.cpp (hunk 1 above)
Comment (ldionne) [reply libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/abi.trivially_copyable.compile.pass.cpp:40]: I have no idea, removed.

**Pull request merged**